Refactor optimize

andron opened the pull request:
Refactoring of visitors in optimize.c to simplify (perhaps) and remove boiler plate code. A lot of the visitors uses the exact same code pattern to optimize and check for errors.

In commit baef3a1 a fail compilation test fails but the change should provide a more correct solution. Both lhs and rhs should be checked for errors in a "foo || bar" expression.

andron committed:
[refactor] Use optimize method for UnaExp and BinExp visitors

Instead of having boiler plate code in visit() methods replace with a
method (optimizeChildren). Adds error checking in CatExp for e1 and e2,
i.e. check if e1 and e2 are != TOKerror which was not done before.

Update to runnable test case "constfold.d" to capture expression of type
```x && 1/0``` and ```x || 1/0``` that shall ignore rhs when x is
compile time constant false/true respectively.

andralex commented:
@yebblies @WalterBright comments after the update?

andralex commented:
FWIW "85 additions and 140 deletions" looks like a step forward to me. Love the red.

yebblies commented:
Running optimize on both `e1` and `e2` of `&&`/`||` will cause this code to error.

```d
void main()
{
    enum x = 0;
    int y = x && 1 / x != 7;
}
```

andralex commented:
back to @andron

andron commented:
Ok, I see the problem. There should be a test that catches this construct.
```
void main() {
  enum x = 0;
  int y = x && 1 / x;
}
```
This should be enough and the test shall compile. With y = 0.

Though I wonder from a more philosophical point, is the code above correct or incorrect? If you can detect a compile time division-by-zero, isn't that code incorrect? Should the RHS of operator && be evaluated if the LHS is identical to zero at compile time?

A well, there are probably more tricky cases. I'll look into the && and || operators and see if I can get a testcase capturing yebblies example.

yebblies commented:
> Though I wonder from a more philosophical point, is the code above correct or incorrect? If you can detect a compile time division-by-zero, isn't that code incorrect?

The division-by-zero will never happen, so it's not really incorrect.

> Should the RHS of operator && be evaluated if the LHS is identical to zero at compile time?

Obviously not, because `&&` short-circuits.

IMO the deciding factor is that one behavior already exists and should therefore be preserved, unless we have a very good reason to change it.

andron commented:
Yes I realize that the ```x && y``` will break a lot of stuff if y where to be evaluated :). For example the c++ iterator construct ```if (it != v.end() && it->compute()) { ...``` or something similar is used now and then in the code I work with. Though it is run-time so I thought finding dangerous constructs at compile time could be nice. 

I reverted the changes to AndAndExp and OrOrExp, and added two test cases which should help (at least me) in future. I'll see if I can figure out the unreachable thing. I suppose that will issue a warning so that the code can be looked into.

yebblies commented:
> In my opinion, there is no question about whether the code is correct due to short-circuit evaluation. Whether the RHS of && should be evaluated is not up for discussion, the language spec mandates that it must not.

It's not quite that clear-cut IMO.  In this code the division is also never meant to be evaluated, yet you still get an error from the const-folder about it.

```d
void main()
{
    enum x = 0;
    if (x)
        int y = 1 / x != 7;
}
```

andron commented:
I know to little about the language spec for D, but should if-statements also have the short-circuit evalation property? Btw, I get the "divide by 0" error twice when compiling the code above.

yebblies commented:
That's the difference between evaluation and const-folding.  Const-folding may still happen on unreachable code, and you can still get errors from it.  For `&&`/`||` const-folding follows the same rules as evaluation, but we have no tracking of unreachable statements for const-folding.

> Btw, I get the "divide by 0" error twice when compiling the code above, using the master branch.

You really shouldn't.  I don't.

andron commented:
The behaviour should be the same as before after the revert now at least. Adding trackning of unreachable statements should probably be done in a separate PR.

And I solved the twice-error. I use ccache and was switching between branches and that did not trigger a proper rebuild I suppose. A full distclean solved it.

yebblies commented:
Ok, looks good.  Please put both test cases into `runnable/constfold.d` and squash all the commits together and I'll merge.

andron commented:
Great! All squashed.

yebblies commented:
Still need to delete `compilable/test3697and` and `compilable/test3697or`, there's no need to have them twice.

andron commented:
Ah, sloppy of me. Have had a busy weekend. Now the files are gone!

yebblies commented:
Auto-merge toggled on
